Implement QPromise<T>::convert<U>()

dpurgin opened the pull request:
Hi Simon, 

I've implemented an initial version of `QPromise<T>::as<U>()` for casting arithmetic types, conversion to `void`, conversion of QVariant and copy-ctor calls for user-defined types.

Could you take a look please? I'll add the documentation and examples as soon as the implementation is done

Cheers

simonbrunel reviewed (changes requested):
Thanks @dpurgin, it looks good.

I'm wondering if it makes sense to have 2 methods:
- `QPromise::convert<U>()`: the behavior implemented in this PR.
- `QPromise::cast<U>() or ::as<U>()`: only `static_cast<U>` without trying to convert the original value. No `QVariant` involved and it would not compile if the cast is not possible. I think we could optimize it by casting the internal shared pointer only.

I'm really not sure which approach is better (1 vs  2 methods), what do you think?

*nit:* if it's not already the case, make sure to run clang-format.

dpurgin commented on src/qtpromise/qpromisehelpers_p.h:
```diff
@@ -88,6 +88,61 @@ void connectDestroyedToReject(const QtPromise::QPromiseConnections& connections,
 }
 
+template<typename T, typename U>
+struct AreArithmeticTypes
```
True, we can possibly decide for `static_cast`. My original idea was to call the constructor To(From) for non-arithmetic types to allow conversions like QByteArray -> QString. Initially, I thought there should be a distinction: if we decided for uniform initialization like in `To{value}`, we would run into a problem with `double` to `int` conversion because it's a narrowing conversion, and the uniform initialization generates a warning [1]. As for static cast, I thought you had to define a conversion operator in user-defined classes, but apparently a non-explicit constructor with a single argument of the required type is called a _converting constructor_ and is good enough ([2, 3, 4]).

So we can stick to `static_cast`.

[1] https://godbolt.org/z/xKMdxW
[2] https://en.cppreference.com/w/cpp/language/cast_operator
[3] https://en.cppreference.com/w/cpp/language/converting_constructor
[4] https://godbolt.org/z/8GdKEG

simonbrunel replied:
Thank you for the detailed answer, let's use `static_cast` and get rid of the special case.

dpurgin commented on tests/auto/qtpromise/qpromise/tst_as.cpp:
```diff
@@ -0,0 +1,103 @@
+    void fulfillQVariantAsInt();
+    void fulfillQVariantQStringAsInt();
+    void fulfillQByteArrayAsQString();
+
```
Thanks, will do. I wanted to let you verify the API/implementation before adding more tests

dpurgin commented on tests/auto/qtpromise/qpromise/tst_as.cpp:
```diff
@@ -0,0 +1,103 @@
+
+void tst_qpromise_as::rejectUnconvertibleQVariant()
+{
+    auto p = QtPromise::resolve(QVariant{"42foo"}).as<int>();
```
Hm, I think QVariant<T> -> QVariant<U> will always work for basic types because QVariant is "intelligent" enough to perform casts, and the primitive types that are valid for QVariant are all arithmetic, therefore convertible to each other (possibly with loss of precision, but still) [1]

[1] https://en.cppreference.com/w/cpp/language/type

simonbrunel replied:
Good point.

dpurgin commented on tests/auto/qtpromise/qpromise/tst_as.cpp:
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Simon Brunel, https://github.com/simonbrunel
+ * Copyright (c) Dmitriy Purgin, https://github.com/dpurgin
```
I've modified a substantial portion of the file (the 'substantiality' is of course subjective, but, in fact, this file has been added by me), so the line refers to this file. Would you prefer another format to address the others? Like one line 'Simon Brunel and contributors'?

simonbrunel replied:
I understand why you added it only in this file, but what does that mean legally at the file level but also at the project level?

Anyway, I don't want to have to maintain different copyright in each file and keep track of who created a file or who changed most of it. The contribution history is available in git and easily accessible in many places in GitHub ([contributors](https://github.com/simonbrunel/qtpromise/graphs/contributors), [blame](https://github.com/simonbrunel/qtpromise/blame/master/src/qtpromise/qpromise_p.h), etc.). I also want the same header in all files but I'm not sure yet how I want to change it. Let me think about it, however I would like to make this change in a separate commit outside this PR.

dpurgin commented on src/qtpromise/qpromisehelpers_p.h:
```diff
@@ -88,6 +88,61 @@ void connectDestroyedToReject(const QtPromise::QPromiseConnections& connections,
+    : std::integral_constant<bool, std::is_arithmetic<T>::value && std::is_arithmetic<U>::value>
+{ };
+
+template<typename From, typename To, bool = AreArithmeticTypes<From, To>::value>
```
Sure, will do

dpurgin commented on src/qtpromise/qpromisehelpers_p.h:
```diff
@@ -88,6 +88,61 @@ void connectDestroyedToReject(const QtPromise::QPromiseConnections& connections,
+{ };
+
+template<typename From, typename To, bool = AreArithmeticTypes<From, To>::value>
+struct PromiseConverter;
```
I'd move the declarations to qpromise_p.h and the implementations to qpromise.inl since the QVariant->U convertor throws an exception defined in qpromiseexceptions.h. Is this okay or should there be a `qpromise_p.inl`?

simonbrunel replied:
Having `PromiseConverter` in a new file will be inconsistent with other internal classes. And at some point, I'm thinking to merge `qpromise.inl` in `qpromise.h` to be consistent with Qt. So I'm a bit reluctant to introduce a new `qpromise_p.inl` file.

I think logic in `qpromise_p.h` should be able to throw QPromise* exceptions. I'm wondering if moving `QPromiseError` (deprecated) at the end of `qpromise_p.h`  (so we can remove the `qpromise_p.h` include) would be a breaking change (if someone included `qpromiseexceptions.h` directly)?

dpurgin replied:
I've started implementing your feedback today, and the problem is not only in QPromise* exceptions but it is actually the implementations of the ` PromiseConverter::call()` themselves: they require the return type `QPromise<T>`  to be complete at this point. So I've ended up puttint the ` PromiseConverter`  implementations at the end of ` qpromise.h` so far. 

Regarding `QPromiseError`: is `qpromiseexceptions.h` supposed to be included directly? I think there could be a preprocessor check in `qpromiseexceptions.h` to verify whether `qpromise_p.h` has been included before qpromiseexceptions.h or not. If not, then we know that somebody has included `qpromsieexceptions.h`  directly, and we could emit a compiler error saying to update the includes, something like:

```c++
// qpromisexceptions.h

#if !defined(QTPROMISE_QPROMISE_P_H)
#error Do not use qpromiseexceptions.h, include QtPromise instead.
#endif
```

It's not really elegant, but it should work

simonbrunel replied:
> ...is qpromiseexceptions.h supposed to be included directly?

Yes, all files that don't have the `_p.h` suffix can be included directly. Though, as written in the comment, `QPromiseError` was not documented and wasn't supposed to be used outside the lib and since it's unlikely someone included `qpromiseexceptions.h` directly,I would not try to make this case non breaking. I will mention it in the release notes.

> ...they require the return type QPromise<T> to be complete at this point.

Could we return the lambda instead of passing the promise to the converter:

```c
template<typename T, typename U>
struct PromiseConverter
{
    // NOTE: can we use `auto` instead of `std::function`?
    static std::function<U(const T&)> create()
    {
        return [](const T& value) {
            return static_cast<U>(value);
        };
    }
};

// ...

template<typename T>
struct PromiseConverter<T, void>
{
    static std::function<void()> create()
    {
        return []() {};
    };
};

// ...

template<typename T>
template<typename U>
inline QPromise<U> QPromiseBase<T>::as() const
{
  return then(QtPromisePrivate::PromiseConverter<T, U>::create());
}
```

dpurgin replied:
> Could we return the lambda instead 
Oh wow, that's a neat trick! Worked like a charm! I've been able to put the converters into `qpromise_p.h` now, thanks

dpurgin commented:
Hey Simon, thanks for the review. 

My Qt Creator runs clang-format automatically upon editing, so this shouldn't be a problem, but I'll make sure I reformat my changes again. 

As for two separate methods `convert` and `cast/as`, I'm a bit unsure. On one side, this makes clear that the intention is to convert a variant (we could also add `std::variant` or `std::any`) and to cast otherwise. On the other hand, since both words are kind of synonyms, this could lead to a confusion: one would have to remember, what the difference between `convert` and `cast` is, whether `cast` is dynamic, static, or reinterpret. In the original problem where I wanted to just drop the fulfilled value and convert to `QPromise<void>`, I think I would be wondering whether I should call `convert` or `cast` if I'm a new user.

Generally, I support the separation of the functionality into `convert` and `cast` but I also think it could be too much of an API for a simple procedure. We could probably implement two methods and deprecate one of them if it causes a lot of questions.

simonbrunel commented:
I'm not sure there is a use case for calling the *cast* method instead of the *conversion* one. And I agree that one method should be enough and easier to understand. If someone really wants to avoid calling the conversion API, it's still easy to write:

```cpp
p.then([](T& value) { 
    return static_cast<U>(value)
});
```

dpurgin commented:
Simon, could you please review the latest update: 

* Converters are moved to `qpromise_p.h`, their implementations return a convertion lambda
* Support for `void`<->`QVariant`
* `U{T}` removed, `static_cast<U>(T)` is the primary converter
* Copyright header in `tst_as.cpp` is the same as in all other files
* Some more tests added to cover all convertors.

I'm not sure about testing the other QVariant conversions with built-in Qt types like in your branch. They all call the same converter and, ultimately, we will be testing the class QVariant instead of QtPromise. I understand that formally it is a black box testing procedure and all these tests would make sense, but then there is a question how many Qt types should be tested, how do we pick them and where to stop? 

If the behavior and the API of the PR are ok, I'll start with the docs.

Cheers

simonbrunel reviewed (changes requested):
> They all call the same converter and, ultimately, we will be testing the class QVariant instead of QtPromise.

I agree, current tests should be enough. If a bug for a specific type is reported, then we will add a regression test for it. Though, I will need to take another look at your tests since I don't have time to review it now.

> If the behavior and the API of the PR are ok, I'll start with the docs.

I left a comment about `void -> QVariant` which should be the only one to impact the docs.

Thanks @dpurgin

dpurgin commented on tests/auto/qtpromise/qpromise/tst_as.cpp:
```diff
@@ -0,0 +1,184 @@
+struct Foo
+{
+    Foo() = default;
+    Foo(int bar_) : bar{bar_} { }
```
Do you have a preference regarding struct or class and member exposure directly or or via accessors in a unit test setting? i.e.: 
```c++
struct Foo 
{ 
  int m_bar; 
};
``` 
or
```c++
class Foo 
{ 
public: 
  int bar(); 
private: 
  int m_bar; 
};
```

simonbrunel replied:
I prefer accessors when the class is shared, else, if it's internal to a specific test files, I don't have a preference.

dpurgin commented on src/qtpromise/qpromise_p.h:
```diff
@@ -553,6 +556,76 @@ struct PromiseInspect
+struct PromiseConverter<
+    T,
+    U,
+    typename std::enable_if<!std::is_same<T, void>::value && !std::is_same<U, void>::value
```
Very true, `std::enable_if` is not pretty but that's what we have. I although thought about making some dispatching like 

```c++
template<typename T, 
  IsQVariant = std::is_same_v<T, QVariant>, 
  IsStdVariant = std::is_same_v<T, std::variant>,
  /* and so on */ >
struct ConverterTypeTraits {
};

template<typename T>
using CastableT = ConverterTypeTraits<T, false, false>;

template<typename T>
using QVariantT = ConverterTypeTraits<T, true, false>;

template<typename T>
using StdVariantT = ConverterTypeTraits<T, false, true>;

template<typename T, typename U>
struct PromiseConverter<T, U, CastableT>
{ 
     // static_cast
};

template<QVariant, typename U>
struct PromiseConverter<T, U, QVariantT>
{
    // QVariant convert
};

template<typename U>
QPromise<T>::as()
{
    then(PromiseConverter<T, ConverterTypeTraits<U>>::create());
}
```

I'm not sure that would work though and whether it is not actually more boilerplate than there is now. We could get back to this approach again once there will be more types to dispatch

dpurgin commented on src/qtpromise/qpromise_p.h:
```diff
@@ -553,6 +556,76 @@ struct PromiseInspect
+    static std::function<QVariant()> create()
+    {
+        return []() {
+            return QVariant{};
```
The only use case for casting void -> U (including QVariant) I can think of is a processing of differently typed promises in a construction like:

```
makePromiseList({QPromise<void>, QPromise<int>, QPromise<QVariant>})
.as<QString>()
.all([](const QString& val) {
  // output is QString{}, probably a QString::number(), and QVariant::toString()
});
```

Since we don't have this kind of promises list, I can't think of any other use case for such a conversion. I think we can safely drop it for now and reimplement rather quickly if there is a feature request.

`thenReturn` is an interesting approach but looks like a technicality that would require some precaution in C++, right?  If it's not a literal that is returned, then such a variable must be able to outlive the scope where the promise was created, so it's either a raw pointer, or some kind of a smart pointer captured in both lambdas. 

Anyway, for me `p.thenReturn()` sounds like there cannot be any more promise chaining after this. On the other hand, those who are familiar with bluebird might find it useful. Generally, I'm for increasing recall and recognition, so there might be a `thenReturn` in QtPromise as well.

dpurgin replied:
By removing the `void -> QVariant` cast I've got rid of `std::enable_if` completely

simonbrunel replied:
`thenReturn()` would not be hard to implement since it would perform a copy (or move) of the returned value. It would also support chaining since it returns a promise:

```cpp
QPromise<T>::thenReturn(V value) const
{
    auto p = QtPromise::resolve(std::forward<V>(value));
    return then([=]() { return p; });
}

// void specialization:
QPromise<void>::thenReturn() const
{
    return then([=]() { });
}
```

dpurgin commented on src/qtpromise/qpromise_p.h:
```diff
@@ -553,6 +556,76 @@ struct PromiseInspect
+};
+
+template<typename T>
+struct PromiseConverter<T, void>
```
True, [1] says that `static_cast<void>` discards the expression after evaluating it. Since we don't really have an expression with side effects, I've removed the U=void specialization

[1] https://en.cppreference.com/w/cpp/language/static_cast

dpurgin commented:
Hey Simon, could you review the last changes please

* docs added
* `void -> QVariant`  conversion removed, `std::enable_if` removed as well
* `PromiseConverter<T, void>`  removed

Cheers

simonbrunel reviewed (changes requested):
Really good work! 

Most of my comments are minor changes.

While reading the docs, I feel that `.as()` is a mix of the `.cast()` vs `.convert()` APIs and I start to think it provides an inconsistent behavior because it allows to convert `QVariant(string) -> int` but not `string -> int` (and vice-versa). 

Alternative behaviors would be:
1. don't support converting `QVariant` at all (so only rely on static_cast).
2. only convert `QVariant` if the original value is of the same type as `U` (not sure it's possible).
3. or support the same conversion logic as `QVariant`.

(1) looks the most consistent. I don't see a good use case for (2) while (3) could be interesting as a real conversion API. Though, the current implementation is probably not an issue, we can still split it in `.cast` & `.convert` later (and deprecate `.as`) if it's confusing / problematic to some users.

Thoughts?

simonbrunel commented:
**Solution 4:** have an explicit API that converts the value as described in (3) and support simple `static_cast` via assignment operator / copy constructors:

```cpp
// static_cast
QPromise<int> p0 = QtPromise::resolve(42);
QPromise<double> p1 = p0;
QPromise<void> p2 = p1;
QPromise<void> p3 = QtPromise::all<void>({p0, p1, p2});

// QPromise<QString> p4 = p0; // FAIL

// convert
QPromise<QString> p5 = p0.as<QString>();
```

dpurgin commented:
Hey @simonbrunel, 

could you please review the changes as discussed on Slack: 
* `as<U>()`  renamed to `convert<U>()` and moved from `QPromiseBase<T>` to `QPromise<T>`
* Added a "generic" conversion via `QVariant`

I've specialized the `static_cast` specialization not only for arithmetic types, but also for types having a converting constructor and for enums as well. Unfortunately, there is no one single compile-time check to flag such types, so I've ended up with a series of conditions. On the bright side, I've managed to dispatch it without the `enable_if` stuff.

Cheers

simonbrunel commented on tests/auto/qtpromise/qpromise/tst_convert.cpp:
```diff
@@ -0,0 +1,257 @@
+    QVERIFY(p.isFulfilled());
+}
+
+void tst_qpromise_convert::rejectUnconvertibleQVariant()
```
We should probably test rejecting for non Qt types, for example: `std::vector` to `int`?

dpurgin replied:
Good point, I've added a test for it

simonbrunel reviewed (commented):
Thank you @dpurgin, really good work!

I didn't know about `QMetaType::registerConverter`, that's really great to extend the QPromise conversion API. That could be also something to investigate in order to implicitly convert a QPromise to a QML promise (and vice-versa) (i.e. related to #5 / [this branch](https://github.com/simonbrunel/qtpromise/tree/wip/qtqmlpromise)).

Indeed, nice trick to avoid duplicating the `std::enable_if` logic. I'm still not fan of the code layout it produces, but I can figure out later how to make it looks nicer - or just live with it :)

I left a few comments, all minors.

dpurgin commented on docs/qtpromise/qpromise/convert.md:
````diff
@@ -0,0 +1,118 @@
+});
+```
+
+If `U` has a [converting constructor](https://en.cppreference.com/w/cpp/language/converting_constructor) 
````
At first, I thought of `std::vector` accepting the size, but turns out this constructor is explicit: `https://en.cppreference.com/w/cpp/container/vector/vector`. 

Of course, this can happen if there is a constructor satisfying the converting constructor properties but not performing a conversion, and it probably might happen to some user classes (I think not everybody is strict with `explicit`). I believe for library classes we should trust the library developers to have produced the correct class definition, and for user classes we can't do anything except mentioning it in the docs. Anyway, I'm confident that QtPromise users are proficient enough to not expect some magical conversion to happen to their classes not having implemented it specifically in code :-)

I've elaborated in the docs on what a converting constructor is (because one extra click on the link is usually too many) and added a note in the docs regarding the correct behavior of the converting constructor.

simonbrunel replied:
> because one extra click on the link is usually too many

Yes and it's also easier to understand the way you described it in the docs.

Thanks!

dpurgin commented:
Hi @simonbrunel, 

thanks for the comments! I've fixed them in the recent commits. 

Cheers

simonbrunel commented:
Thank you @dpurgin for your work!
